File: src/python/xerxes/extensions/skill_sources/official.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Xerxes-curated official skill registry (HTTP-fetched)."""

from __future__ import annotations

import json
from typing import Any

import httpx

from .base import SkillBundle, SkillSearchHit, SkillSource


class OfficialSkillSource(SkillSource):
    """HTTP-backed skill source for the Xerxes-curated registry."""

    name = "official"

    def __init__(self, *, base_url: str = "https://skills.xerxes-agent.dev", client: httpx.Client | None = None) -> None:
        """Bind the source to ``base_url`` and optionally an existing client.

        Args:
            base_url: Base URL of the registry.
            client: Optional pre-configured ``httpx.Client`` for testing.
        """
        self._base_url = base_url
        self._client = client

    def _http(self) -> httpx.Client:
        """Return the injected client or build a fresh ``httpx.Client``."""
        return self._client or httpx.Client(base_url=self._base_url, timeout=20.0)

    def search(self, query: str, *, limit: int = 20) -> list[SkillSearchHit]:
        """Query the ``/search`` endpoint and return at most ``limit`` hits."""
        with self._http() as c:
            try:
                resp = c.get("/search", params={"q": query, "limit": limit})
                resp.raise_for_status()
            except httpx.HTTPError:
                return []
            try:
                rows: list[dict[str, Any]] = resp.json()
            except json.JSONDecodeError:
                return []
        return [
            SkillSearchHit(
                name=r["name"],
                description=r.get("description", ""),
                source_name=self.name,
                version=r.get("version", ""),
                tags=list(r.get("tags", [])),
            )
            for r in rows
        ]

    def fetch(self, identifier: str) -> SkillBundle:
        """Download the ``SKILL.md`` for ``identifier`` from the registry."""
        with self._http() as c:
            resp = c.get(f"/skills/{identifier}/SKILL.md")
            resp.raise_for_status()
            body = resp.text
        return SkillBundle(name=identifier, version="official", body_markdown=body, source_name=self.name)


__all__ = ["OfficialSkillSource"]

File: src/typescript/src/extensions/skillSources/official.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  normalizeSkillSearchLimit,
  requireSkillIdentifier,
  requireSkillSearchQuery,
  SkillSourceError,
  type SkillBundle,
  type SkillSearchHit,
  type SkillSource,
} from './base.js'

export const DEFAULT_OFFICIAL_SKILL_REGISTRY_URL = 'https://skills.xerxes-agent.dev'

/** The minimal response shape the source needs from a host-approved fetch implementation. */
export interface SkillRegistryFetchResponse {
  readonly ok: boolean
  readonly status: number
  text(): Promise<string>
}

/** Explicit remote boundary. The source never falls back to global `fetch`. */
export interface SkillRegistryFetchTransport {
  fetch(url: string, init: Readonly<{ method: 'GET' }>): Promise<SkillRegistryFetchResponse>
}

export interface OfficialSkillSourceOptions {
  readonly baseUrl?: string
  readonly transport: SkillRegistryFetchTransport
}

/** Raised for a non-successful registry download rather than manufacturing an empty bundle. */
export class SkillRegistryHttpError extends SkillSourceError {
  readonly status: number

  constructor(url: string, status: number) {
    super(`skill registry request failed with HTTP ${status}: ${url}`)
    this.name = 'SkillRegistryHttpError'
    this.status = status
  }
}

/** HTTP-backed official catalogue whose transport is supplied by the embedding application. */
export class OfficialSkillSource implements SkillSource {
  readonly baseUrl: string
  readonly name: string = 'official'
  protected readonly transport: SkillRegistryFetchTransport

  constructor(options: OfficialSkillSourceOptions) {
    this.baseUrl = normalizeRegistryBaseUrl(options.baseUrl ?? DEFAULT_OFFICIAL_SKILL_REGISTRY_URL)
    this.transport = options.transport
  }

  async search(query: string, limit = 20): Promise<readonly SkillSearchHit[]> {
    const normalizedQuery = requireSkillSearchQuery(query)
    const normalizedLimit = normalizeSkillSearchLimit(limit)
    if (normalizedLimit === 0) return []
    const url = registryUrl(this.baseUrl, 'search', { q: normalizedQuery, limit: String(normalizedLimit) })
    try {
      const response = await this.transport.fetch(url, { method: 'GET' })
      if (!response.ok) return []
      const rows = parseSearchRows(await response.text())
      return rows.slice(0, normalizedLimit).map(row => ({
        name: row.name,
        description: row.description,
        sourceName: this.name,
        version: row.version,
        tags: row.tags,
      }))
    } catch {
      // Search is best-effort by design, matching the Python registry source.
      return []
    }
  }

  async fetch(identifier: string): Promise<SkillBundle> {
    const id = requireSkillIdentifier(identifier)
    const url = registryUrl(this.baseUrl, `skills/${encodeURIComponent(id)}/SKILL.md`)
    const response = await this.transport.fetch(url, { method: 'GET' })
    if (!response.ok) throw new SkillRegistryHttpError(url, response.status)
    return {
      name: id,
      version: 'official',
      bodyMarkdown: await response.text(),
      metadata: {},
      sourceName: this.name,
    }
  }
}

interface RegistrySearchRow {
  readonly description: string
  readonly name: string
  readonly tags: readonly string[]
  readonly version: string
}

function normalizeRegistryBaseUrl(value: string): string {
  try {
    const url = new URL(value)
    if (url.protocol !== 'http:' && url.protocol !== 'https:') {
      throw new SkillSourceError('skill registry baseUrl must use http or https')
    }
    url.hash = ''
    url.search = ''
    url.pathname = `${url.pathname.replace(/\/+$/, '')}/`
    return url.toString()
  } catch (error) {
    if (error instanceof SkillSourceError) throw error
    throw new SkillSourceError(`invalid skill registry baseUrl: ${value}`, { cause: error })
  }
}

function registryUrl(baseUrl: string, path: string, query: Readonly<Record<string, string>> = {}): string {
  const url = new URL(path, baseUrl)
  for (const [key, value] of Object.entries(query)) url.searchParams.set(key, value)
  return url.toString()
}

function parseSearchRows(body: string): readonly RegistrySearchRow[] {
  const value: unknown = JSON.parse(body)
  if (!Array.isArray(value)) throw new SkillSourceError('skill registry search response must be a JSON array')
  return value.map(row => {
    if (!isRecord(row) || typeof row.name !== 'string' || !row.name.trim()) {
      throw new SkillSourceError('skill registry search row must include a skill name')
    }
    return {
      name: row.name,
      description: typeof row.description === 'string' ? row.description : '',
      version: typeof row.version === 'string' ? row.version : '',
      tags: Array.isArray(row.tags) && row.tags.every(tag => typeof tag === 'string') ? [...row.tags] : [],
    }
  })
}

function isRecord(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}
